scalableminds/webknossos
# Add drag&drop nml upload functionality to dashboard

This PR adds the possibility to drag&drop (multiple) nml or zip files onto the dashboard to upload them. I reused the Dropzone that is also used in the tracing view (very easy to reuse, nice work @philippotto) :)

@fm3 A flag called `createGroupForEachFile` with a string value of `"true"` or `"false"` is now sent as part of the multipart request to `/api/annotations/upload` as well. This flag indicates whether the trees of each nml file should be put into separate tree groups, or not. To wrap all trees of an nml file in a new group, create a new group with an unused groupId and assign the original tree groups as children of that new group. The name of the new group should be the name of the nml file. Additionally assign the new groupId to all trees that do not have a groupId yet. This should do the trick :)
Have a look at the `wrapInNewGroup` function in the `nml_helpers.js` if you want to see the frontend code to do this.

@philippotto After @fm3 implemented the backend part, this should be ready to test and review.

### URL of deployed dev instance (used for testing):
- https://dnddashboard.webknossos.xyz

### Steps to test:
- Drag&Drop (multiple) nml or (a single) zip file(s) to the dashboard (the activated tab shouldn't be relevant). A modal should appear that shows some basic information about those files.
- Try to upload with and without the `createGroupForEachFile` option set. The uploaded annotations should be merged, you should be redirected to the tracing view and see the merged annotation. The trees should be in separate groups or not, according to the option you set.
- The Drag&Drop upload functionality in the tracing view should work in the same way as before.
- Try to upload a faulty NML. An error message should be displayed and the upload modal should be dismissed.

### Issues:
- fixes #2926
- fixes #2972

------
